docs/knn/index.md

```diff
@@ -246,3 +246,98 @@ All parameters are required.
+
+
+## Use Similarity methods in painless scripting
+
+Sometimes users would like to go beyond Elasticsearch’s built-in features for scoring and might want to customize the search scores in more complex ways.
+Elasticsearch provides script_score, the ability to provide custom scores for desired documents.
+Elasticsearch Painless, a simple, secure scripting language designed specifically for this purpose.
+
+A Painless script is evaluated within a context. Painless has a strict list of allowed methods and classes per context to ensure all Painless scripts are secure.
+Now, opendistro users can use following specialized methods within Score context to customize their scores.
+
+### Cosine Similarity
+This function calculates the measure of cosine similarity between a given query vector and document vectors.
+Optionally accepts normQueryVector, to avoid repeated calculation of normalization for query vector for every filtered documents.  
+#### Definition
+```
+float cosineSimilarity (float[] queryVector, doc['vector field']) 
+float cosineSimilarity (float[] queryVector, doc['vector field'], float normQueryVector) 
+```
+
+#### Usage
+```
+GET my-knn-index-2/_search
+{
+  "size": 2,
+  "query": {
+    "script_score": {
+      "query": {
+        "bool": {
+          "filter": {
+            "term": {
+              "color": "BLUE"
+            }
+          }
+        }
+      },
+      "script": {
+        "source": "1.0 + cosineSimilarity(params.query_value, doc[params.field])", 
+        "params": {
+          "field": "my_vector",
+          "query_value": [9.9, 9.9],
+        }
+      }
+    }
+  }
+}
+```
+The above script adds 1.0 to the cosine similarity to keep score positive.
+### L2 Squared
+This function calculates L2 distance (Euclidean distance) between a given query vector and document vectors.
+
+#### Definition
+```
+float l2Squared (float[] queryVector, doc['vector field'])
+```
+#### Usage
+```
+GET my-knn-index-2/_search
+{
+  "size": 2,
+  "query": {
+    "script_score": {
+      "query": {
+        "bool": {
+          "filter": {
+            "term": {
+              "color": "BLUE"
+            }
+          }
+        }
+      },
+      "script": {
+        "source": "1 / (1 + l2Squared(params.query_value, doc[params.field]))", 
+        "params": {
+          "field": "my_vector",
+          "query_value": [9.9, 9.9],
+        }
+      }
+    }
+  }
+}
+
+```
+The lesser the distance the more the relevance of the document to the query vector. In order to bring the lesser 
+distances documents to the top of the scores, we invert the distance from l2Squared function.
+Also, when a document vector matches the query vector, we needed to add 1 in the denominator to avoid divide by zero error.
+
+####Constraints
+1. If a document’s knn vector field has different dimensions from the query, an error(IllegalArgumentException) will be thrown.
```

Comment:
1. If a document’s `knn_vector` field has different dimensions than the query, the function throws an IllegalArgumentException.
2. If a vector field doesn't have a value, the function throws an IllegalStateException.

   You can avoid this situation by first checking if a document as a value for the field:

   ```
   "source": "doc[params.field].size() == 0 ? 0 : 1 / (1 + l2Squared(params.query_value, doc[params.field]))",
   ```

   Since scores can only be positive, this script ranks documents with vector fields higher than those without.

Reply:
Ack